guide/src/async-await.md

```diff
@@ -0,0 +1,240 @@
+    pyo3_asyncio::async_std::into_coroutine(py, async move {
+        rust_sleep().await;
+        Ok(())
+    })
+}
+```
+
+In Python, we can call this pyo3 function just like any other async function:
+
+```python
+from example import call_rust_sleep
+
+async def rust_sleep():
+    await call_rust_sleep()
+```
+
+## Managing Event Loops
+
+Python's event loop requires some special treatment, especially regarding the main thread. Some of
+Python's Asyncio features, like proper signal handling, require control over the main thread, which
+doesn't always play well with Rust.
+
+Luckily, Rust's event loops are pretty flexible and don't _need_ control over the main thread, so in
+pyo3-asyncio, we decided the best way to handle Rust/Python interop was to just surrender the main
+thread to Python and run Rust's event loops in the background. Unfortunately, since most event loop 
+implementations _prefer_ control over the main thread, this can still make some things awkward.
+
+### PyO3 Asyncio Initialization
+
+Because Python needs to control the main thread, we can't use the convenient proc macros from Rust
+runtimes to handle the main function or test functions. 
+
+Instead, the initialization for PyO3 has to be done manually from the main function and the main 
+thread must block on `pyo3_asyncio::run_forever` or `pyo3_asyncio::generic::run_until_complete`.
+Because we have to block on one of those functions, we can't use `tokio::main` since it's not a good
+idea to make long blocking calls during an async function.
+
+In addition, some runtimes, such as Tokio, may require some additional initialization since their 
+runtimes are customizable. For tokio, this initialization happens during the `tokio::main` proc 
+macro, but since we can't use that for our purposes, it has to be initialized manually. See the 
+`pyo3-asyncio` API docs for more information.
+
+Here's a full example of PyO3 initialization:
+```rust
+use pyo3::prelude::*;
+
+fn main() {
+    // if using tokio, you should perform some additional initialization here:
+    // pyo3_asyncio::tokio::init_multi_thread();
+
+    Python::with_gil(|py| {
+        // Initialize the runtime
+        pyo3_asyncio::with_runtime(py, || {
+            // Run the Python event loop until the given future completes
+            pyo3_asyncio::async_std::run_until_complete(py, async {
+                // PyO3 is initialized - Ready to go
+                Ok(())
+            })?;
+
+            Ok(())
+        })
+        .map_err(|e| {
+            e.print_and_set_sys_last_vars(py);  
+        })
+        .unwrap();
+    })
+}
+```
+
+## PyO3 Asyncio in Cargo Tests
+
+The default Cargo Test harness does not currently allow test crates to provide their own main 
+function, so there doesn't seem to be a good way to allow Python to gain control over the main
+thread.
+
+We can, however, override the default test harness and provide our own. `pyo3-asyncio` provides some
+utilities to help us do just that!
```

Comment:
I would add a quick note here before going into the subsections, saying what we'll need to do (e.g. `Cargo.toml`, custom `pytests` directory, different way to write tests.